src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,269 @@
+end
+
+@generated function (::Type{SSymmetricCompact{N, T, L}})(a::Tuple) where {N, T, L}
+    expr = Vector{Expr}(undef, L)
+    i = 0
+    for col = 1 : N, row = col : N
+        index = N * (col - 1) + row
+        expr[i += 1] = :(a[$index])
+    end
+    quote
+        @_inline_meta
+        @inbounds return SSymmetricCompact{N, T, L}(SVector{L, T}(tuple($(expr...))))
+    end
+end
+
+@inline function (::Type{SSymmetricCompact{N, T}})(a::Tuple) where {N, T}
+    L = triangularnumber(N)
+    SSymmetricCompact{N, T, L}(a)
+end
+
+@inline (::Type{SSymmetricCompact{N}})(a::Tuple) where {N} = SSymmetricCompact{N, promote_tuple_eltype(a)}(a)
+@inline (::Type{SSymmetricCompact{N}})(a::NTuple{M, T}) where {N, T, M} = SSymmetricCompact{N, T}(a)
+@inline SSymmetricCompact(a::StaticMatrix{N, N, T}) where {N, T} = SSymmetricCompact{N, T}(a)
+
+@inline (::Type{SSC})(a::SSymmetricCompact) where {SSC <: SSymmetricCompact} = SSC(a.lowertriangle)
+@inline (::Type{SSC})(a::SSC) where {SSC <: SSymmetricCompact} = a
+
+@inline (::Type{SSC})(a::AbstractVector) where {SSC <: SSymmetricCompact} = SSC(convert(lowertriangletype(SSC), a))
+
+# TODO: similar_type overload?
+
+@generated function _symmetric_compact_indices(::Val{N}) where N
+    # Returns a Tuple{Pair{Int, Bool}} I such that for linear index i,
+    # * I[i][1] is the index into the lowertriangle field of an SSymmetricCompact{N};
+    # * I[i][2] is true iff i is an index into the lower triangle of an N × N matrix.
+    indexmat = Matrix{Pair{Int, Bool}}(undef, N, N)
+    i = 0
+    for col = 1 : N, row = 1 : N
+        indexmat[row, col] = if row >= col
+            (i += 1) => true
+        else
+            indexmat[col, row][1] => false
+        end
+    end
+    quote
+        Base.@_inline_meta
+        return $(tuple(indexmat...))
+    end
+end
+
+Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int) where {N}
+    I = _symmetric_compact_indices(Val(N))
+    j, lower = I[i]
+    @inbounds x′ = a.lowertriangle[j]
+    return lower ? x′ : transpose(x′)
+end
+
+Base.@propagate_inbounds function Base.setindex(a::SSymmetricCompact{N, T, L}, x, i::Int) where {N, T, L}
+    I = _symmetric_compact_indices(Val(N))
+    j, lower = I[i]
+    x′ = lower ? x : transpose(x)
+    return SSymmetricCompact{N}(setindex(a.lowertriangle, x′, j))
+end
+
+# needed because it is used in convert.jl and the generic fallback is slow
+@generated function Base.Tuple(a::SSymmetricCompact{N}) where N
+    exprs = [:(a[$i]) for i = 1 : N^2]
+    quote
+        @_inline_meta
+        tuple($(exprs...))
+    end
+end
+
+LinearAlgebra.ishermitian(a::SSymmetricCompact{N, T}) where {N,T <: Real} = true
+LinearAlgebra.ishermitian(a::SSymmetricCompact) = all(isreal, a.lowertriangle)
+LinearAlgebra.issymmetric(a::SSymmetricCompact) = true
+
+# TODO: factorize?
+
+# TODO: a.lowertriangle == b.lowertriangle is slow (used by SDiagonal). SMatrix etc. actually use AbstractArray fallback (also slow)
```

Comment:
Should/does this have an issue?

Reply:
The speed of the `AbstractArray` fallback seems to no longer be an issue. I'll change this accordingly.